### PR TITLE
Edit Media Item via Hover Dropdown Menu

### PR DESCRIPTION
**Pull request by joannag6**
## Summary

Allow user to edit media items in an edit dialog via the hover dropdown menu.

Screen Shot 2020-05-29 at 3 20 15 PM

Screen Shot 2020-05-29 at 3 20 23 PM

## To-do
Add Uploaded date and File size to the dialog. (Currently this info is not kept in the `resource` object.

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+          <Video key={src} poster={poster} preload="none" muted>
+            <source src={src} type={mimeType} />
+          </Video>
+        )}
+        <MetadataTextContainer>
+          <MediaTitleText>{title}</MediaTitleText>
+          <MediaSizeText>
+            {sprintf(
+              /* translators: %(width)d: image width and %(height)d: image height. */
+              __('%(width)d X %(height)d pixels', 'web-stories'),
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Nothing wrong with this technically, but so far we're not using this form of placeholders AFAIK.
For consistency reasons, I'd recommend using something like this:

```js
sprintf(
  /* translators: 1: image width. 2: image height. */
  __('%d X %d pixels', 'web-stories'),
  resource.width,
  resource.height
)
```

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+/**
+ * Displays a dialog where the user can edit the selected media element.
+ *
+ * @param {Object} props Component props.
+ * @param {Object} props.resource Selected media element's resource object.
+ * @param {Object} props.showEditDialog If dialog should be displayed.
+ * @param {Object} props.setShowEditDialog Callback to toggle dialog display.
+ * @return {null|*} The dialog element.
+ */
+function MediaEditDialog({ resource, showEditDialog, setShowEditDialog }) {
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Might be a nice component to add a storybook story for.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 1 above)
Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Hmm I tried that before and I got this error from eslint:

`Multiple sprintf placeholders should be ordered. Mix of ordered and non-ordered placeholders found.eslint(@wordpress/valid-sprintf)`

Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Oh, my bad.

```js
sprintf(
  /* translators: 1: image width. 2: image height. */
  __('%1$d X %2$d pixels', 'web-stories'),
  resource.width,
  resource.height
)
```

Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Again, your version totally works. I'd just like to keep things a bit consistent 🙂

Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Ok changed it!

File: assets/src/edit-story/app/api/apiProvider.js [hunk 3]
```diff
@@ -194,6 +194,22 @@ function APIProvider({ children }) {
...
+  /**
+   * Delete existing media.
+   *
+   * @param  {number} mediaId
+   * @return {Promise} Media Object Promise.
+   */
+  const deleteMedia = useCallback(
+    (mediaId) => {
+      return apiFetch({
+        path: `${media}/${mediaId}/?force=true`,
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]: See comments here. https://github.com/google/web-stories-wp/pull/1654/files#r422166579

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+`;
+
+const MetadataTextContainer = styled.div`
+  display: flex;
+  flex-direction: column;
+`;
+
+const MediaTitleText = styled.div`
+  font-weight: bold;
+  color: #3c4043;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Colors should not be hardcoded. There should be loaded from theme.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+`;
+
+const MediaSizeText = styled.div`
+  font-size: 16px;
+  color: #3c4043;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Hardcoded color

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+`;
+
+const Input = styled.input`
+  background: #ffffff;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Hardcoded color

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  border: 1px solid #d3d4d4;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Hardcoded color

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+  font-size: 15px;
+  padding: 7px 10px;
+  margin-top: 20px;
+  margin-bottom: 4px;
+`;
+
+const DialogDescription = styled.div`
+  font-size: 12px;
+  line-height: 16px;
+  color: #202124;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Hardcoded color

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    setAltText(evt.target.value);
+  }, []);
+
+  useEffect(() => {
+    const updateMediaItem = async () => {
+      try {
+        await updateMedia(id, { alt_text: altText });
+        setShowEditDialog(false);
+      } catch (err) {
+        // TODO Display error message
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: There is already an existing snackbar message. We should use this here.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 8 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: https://github.com/google/web-stories-wp/blob/08952b335aff7dc906ad6edb5141e916b574cd78/assets/src/edit-story/theme.js#L88

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 6 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: https://github.com/google/web-stories-wp/blob/08952b335aff7dc906ad6edb5141e916b574cd78/assets/src/edit-story/theme.js#L99

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+  padding: 7px 10px;
+  margin-top: 20px;
+  margin-bottom: 4px;
+`;
+
+const DialogDescription = styled.div`
+  font-size: 12px;
+  line-height: 16px;
+  color: #202124;
+  opacity: 0.54;
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Should we use something like this 
https://github.com/google/web-stories-wp/blob/08952b335aff7dc906ad6edb5141e916b574cd78/assets/src/edit-story/theme.js#L125

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+      }
+    };
+    shouldUpdate ? updateMediaItem() : null;
+  }, [altText, id, setShowEditDialog, shouldUpdate, updateMedia]);
+
+  return (
+    <Dialog
+      open={showEditDialog}
+      onClose={() => setShowEditDialog(false)}
+      width={'500px'}
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Are the px need here? Does `width={ 500 }` work?

**Comment by spacedmonkey** [comment]
Screenshot 2020-05-12 at 11 45 16
In my testing, I noticed that a dialog was being created for every element media on the page. I have 500 images in my library, that is 500 div tags to keep track of. I feel like this could be a massive performance issue. Is there no way of having one hidden dialog that is updated with src and alt when made visual?

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+  const [shouldUpdate, setShouldUpdate] = useState(false);
+
+  const handleAltTextChange = useCallback((evt) => {
+    setAltText(evt.target.value);
+  }, []);
+
+  useEffect(() => {
+    const updateMediaItem = async () => {
+      try {
+        await updateMedia(id, { alt_text: altText });
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: The media is not updated in state here. So I didn't see the alt change of the image in the media library. This need to be fixed.

File: assets/src/edit-story/components/library/panes/media/dropDownMenu.js [hunk 13]
```diff
@@ -87,29 +92,37 @@ function DropDownMenu({ showDisplayIcon, menuCallback }) {
...
+          <Popup anchor={moreRef} isOpen={showMenu} width={160}>
+            <DropDownList
+              handleCurrentValue={handleCurrentValue}
+              options={options}
+              toggleOptions={toggleOptions}
+            />
+          </Popup>
+        </>
+      )}
+      <MediaEditDialog
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/dropDownMenu.js]: As noted elsewhere, this should really be one dialog for performance reasons.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+          </Plain>
+          <Plain onClick={() => setShouldUpdate(true)}>
+            {__('Save', 'web-stories')}
+          </Plain>
+        </>
+      }
+    >
+      <DialogBody>
+        {type == 'image' ? (
+          <Image src={src} alt={alt} loading={'lazy'} />
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Which size image is being used here?

**Review by spacedmonkey — DISMISSED** [review]
There must be some kind of testing added here. 

I would also like to see storybook intergration.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  box-sizing: border-box;
+  border-radius: 4px;
+  font-size: 15px;
+  padding: 7px 10px;
+  margin-top: 20px;
+  margin-bottom: 4px;
+`;
+
+const DialogDescription = styled.div`
+  font-size: 12px;
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Ideally all font sizes & line-heights come from the theme

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 14 above)
Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Added a small util (`useSmallestImage`) for this!

File: assets/src/edit-story/components/library/panes/media/useSmallestImage.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * Choose the source URL of the smallest available size image.
+ *
+ * @param {*} resource Image resource object.
+ * @return {string} Source URL of the smallest available size image.
+ */
+function useSmallestImage(resource) {
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/useSmallestImage.js]: Only hooks should start with `use...`, as that's what the linter uses to detect hooks.

Perhaps rename to `getThumbnailUrl` or something.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+
+const DialogDescription = styled.div`
+  font-family: ${({ theme }) => theme.fonts.description.family};
+  line-height: ${({ theme }) => theme.fonts.description.lineHeight};
+  font-weight: ${({ theme }) => theme.fonts.description.weight};
+  font-size: ${({ theme }) => theme.fonts.description.size};
+  color: ${({ theme }) => theme.grayout};
+`;
+
+const imageDialogTitle = __('Edit Image', 'web-stories');
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js:105]: You could use `sprintf` here. So something like this. 
```
/* translators: %s is file type. */
const dialogTitle = sprintf( __('Edit %s', 'web-stories'), type );
/* translators: %s is file type. */
const dialogDescription = sprintf( __(
  'Describe the appearance and function of a %s. Leave empty if the %s is purely decorative',
  'web-stories'
), type );
```
Makes this a little cleaner.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+      showSnackbar({
+        message: __('Failed to update, please try again.', 'web-stories'),
+      });
+    }
+  }, [altText, id, onClose, showSnackbar, updateMedia, updateMediaElement]);
+
+  return (
+    <Dialog
+      open={true}
+      onClose={() => onClose()}
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Why not just simply
```suggestion
      onClose={onClose}
```

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+      title={type == 'image' ? imageDialogTitle : videoDialogTitle}
+      actions={
+        <>
+          <Plain onClick={() => onClose()}>{__('Cancel', 'web-stories')}</Plain>
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Why not?
```suggestion
          <Plain onClick={onClose}>{__('Cancel', 'web-stories')}</Plain>
```

File: assets/src/edit-story/components/library/panes/media/mediaElement.js [hunk 20]
```diff
@@ -231,7 +220,7 @@ const MediaElement = ({
       >
         <Image
           key={src}
-          src={imageSrc}
+          src={useSmallestImage(resource)}
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaElement.js]: React hooks can not be called conditionally.

File: assets/src/edit-story/components/library/panes/media/useSmallestImage.js (hunk 16 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/useSmallestImage.js]: This file also feels like this should like `app > meida > utils` as it acts the media resource object.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+            {type == 'image' ? imageDialogDescription : videoDialogDescription}
+          </DialogDescription>
+        </MetadataTextContainer>
+      </DialogBody>
+    </Dialog>
+  );
+}
+
+MediaEditDialog.propTypes = {
+  resource: PropTypes.object,
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Ain't both of this props required?

**Comment by spacedmonkey** [comment]
This PR didn't work when I was testing it. Even after I enabled the feature flag, the dialog did not appear. Part of review code is seeing how it works in a browser / see what is meant to do. Is this PR meant to work?

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 17 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js:105]: Disagree on that one 🙂 

The sentences might need to be translated slightly different depending on whether it's an image or video (e.g. declension). So not using `sprintf` here is fine.

The second example (`Describe the appearance and function of a %s`) doesn't  work properly in English either:

`Describe the appearance and function of a video` would work
`Describe the appearance and function of a image` would be wrong (should be `an image`).

File: assets/src/edit-story/components/library/panes/media/mediaElement.js (hunk 20 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/mediaElement.js]: It's actually not a hook, see https://github.com/google/web-stories-wp/pull/1733#discussion_r429816020

It needs to be renamed.

File: assets/src/edit-story/components/library/panes/media/useSmallestImage.js (hunk 16 above)
Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/useSmallestImage.js]: Moved to app/media/utils/getThumbnailUrl.js

**Comment by joannag6** [comment]
Hmm, it worked for me, did the More icon show up onhover for you?

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+
+  return (
+    <Dialog
+      open={true}
+      onClose={onClose}
+      title={type == 'image' ? imageDialogTitle : videoDialogTitle}
+      actions={
+        <>
+          <Plain onClick={onClose}>{__('Cancel', 'web-stories')}</Plain>
+          <Plain onClick={updateMediaItem}>{__('Save', 'web-stories')}</Plain>
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: This one should be a primary button, not plain

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+  line-height: ${({ theme }) => theme.fonts.description.lineHeight};
+  font-weight: ${({ theme }) => theme.fonts.description.weight};
+  font-size: ${({ theme }) => theme.fonts.description.size};
+  color: ${({ theme }) => theme.grayout};
+`;
+
+const imageDialogTitle = __('Edit Image', 'web-stories');
+const videoDialogTitle = __('Edit Video', 'web-stories');
+const imageDialogDescription = __(
+  'Describe the appearance and function of an image. Leave empty if the image is purely decorative',
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Add period `.` at the end?

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+  'web-stories'
+);
+const videoDialogDescription = __(
+  'Describe the appearance and function of a video. Leave empty if the video is purely decorative',
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Add period `.` at the end?

Also, should we say "of _the_ video" and "of _the_ image"?

**Review by swissspidy — COMMENTED** [review]
Just a few questions, otherwise looks good

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 22 above)
Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: https://github.com/google/web-stories-wp/pull/1656#discussion_r431052042

**Comment by spacedmonkey** [comment]
This looks good from a code level. Just a small styling issue. 
Screenshot 2020-05-28 at 14 28 17
The buttons look a little weird to me.

File: assets/src/edit-story/components/library/panes/media/mediaEditDialog.js (hunk 22 above)
Comment (joannag6) [reply assets/src/edit-story/components/library/panes/media/mediaEditDialog.js]: Created #2104 for this.

**Comment by joannag6** [comment]
> This looks good from a code level. Just a small styling issue.
> Screenshot 2020-05-28 at 14 28 17
> The buttons look a little weird to me.

Added spacing between the buttons for now. Also opened an issue for this - #2104

**Review by swissspidy — APPROVED** [review]

**Pull request merged**